crates/evm/evm/src/inspectors/revert_diagnostic.rs

```diff
@@ -0,0 +1,207 @@
+    D: Database<Error = DatabaseError>,
+    CTX: ContextTr<Db = D>,
+    CTX::Journal: JournalExt,
+{
+    /// Tracks the first call with non-zero calldata that targets a non-contract address. Excludes
+    /// precompiles and test addresses.
+    fn call(&mut self, ctx: &mut CTX, inputs: &mut CallInputs) -> Option<CallOutcome> {
+        let target = self.code_target_address(inputs);
+
+        if IGNORE.contains(&target) || ctx.journal_ref().precompile_addresses().contains(&target) {
+            return None;
+        }
+
+        if let Ok(state) = ctx.journal().code(target) {
+            if state.is_empty() && !inputs.input.is_empty() {
+                self.non_contract_call = Some((target, inputs.scheme, ctx.journal_ref().depth()));
+            }
+        }
+        None
+    }
+
+    /// Handles `REVERT` and `EXTCODESIZE` opcodes for diagnostics.
+    ///
+    /// When a `REVERT` opcode with zero data size occurs:
+    ///  - if `non_contract_call` was set at the current depth, `handle_revert_diagnostic` is
+    ///    called. Otherwise, it is cleared.
+    ///  - if `non_contract_call` was set at the current depth, `handle_revert_diagnostic` is
+    ///    called. Otherwise, it is cleared.
+    ///
+    /// When an `EXTCODESIZE` opcode occurs:
+    ///  - Optimistically caches the target address and current depth in `non_contract_size_check`,
+    ///    pending later validation.
+    fn step(&mut self, interp: &mut Interpreter, ctx: &mut CTX) {
+        // REVERT (offset, size)
+        if REVERT == interp.bytecode.opcode() {
+            if let Ok(size) = interp.stack.peek(1) {
+                if size.is_zero() {
+                    // Check empty revert with same depth as a non-contract call
+                    if let Some((_, _, depth)) = self.non_contract_call {
+                        if ctx.journal_ref().depth() == depth {
+                            self.handle_revert_diagnostic(interp);
+                        } else {
+                            self.non_contract_call = None;
+                        }
+                        return;
+                    }
+
+                    // Check empty revert with same depth as a non-contract size check
+                    if let Some((_, depth)) = self.non_contract_size_check {
+                        if depth == ctx.journal_ref().depth() {
+                            self.handle_revert_diagnostic(interp);
+                        } else {
+                            self.non_contract_size_check = None;
+                        }
+                    }
+                }
+            }
+        }
+        // EXTCODESIZE (address)
+        else if EXTCODESIZE == interp.bytecode.opcode() {
+            if let Ok(word) = interp.stack.peek(0) {
+                let addr = Address::from_word(word.into());
+                if IGNORE.contains(&addr) ||
+                    ctx.journal_ref().precompile_addresses().contains(&addr)
+                {
+                    return;
```

Comment:
these `return` are incorrect if more checks are added in this function, outside of the `if`; however if you extract to separate functions as mentioned in the other comment these are fine